File: Python Solutions/079 Word Search.py
# -*- coding: utf-8 -*-
"""
Created on Mon Feb 25 10:18:30 2019

@author: Sue

079. Word Search
"""

class Solution(object):
    def __init__(self):
        self.dir = [[-1, 0], # Up
                    [0, 1],  # Right
                    [1, 0],  # Down 
                    [0, -1]  # Left
                    ]
    
    def inArea(self, row, col, x, y):
        if (x >= 0) and (x < row) and (y >= 0) and (y < col):
            return True
        return False
    
    
    def searchAdjacent(self, board, word, visited, x, y):
        """
        :type board: List[List[str]]
        :type word: str : left string to be searched
        :type visited: List[bool] : if the position is visited
        :type x: int
        :type y: int
        :rtype: bool
        """
        
        #All char in word has been found in the board
        if (word == ""):
            return True
        
        #Else, search
        row = len(board)
        if( row == 0 ):
            return False
        col = len(board[0])
                
        if(word[0] == board[x][y]):
            if (len(word[1:]) == 0):
                return True
            
            visited[x][y] = True
            
            #continune searching
            for d in self.dir:
                newX = x + d[0]
                newY = y + d[1]
                
                if (self.inArea(row, col, newX, newY)) and (not visited[newX][newY]):            
                    if (self.searchAdjacent(board, word[1:], visited, newX, newY)):
                        return True
                    else:
                        visited[newX][newY] = False  
                        continue
                    
        else:
            return False
        
        
    
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        #if the word is empty, then return true
        if(word == ""):
            return True
        
        row = len(board)
        if( row == 0 ):
            return False
        
        col = len(board[0])        
            
        for x in range(row):
            for y in range(col):
                visited = [[False] * col for i in range(row)] 
                
                res = self.searchAdjacent(board, word, visited, x, y)
                if (res):
                    return True
                
        return False
        
        

File: CPP Solutions/079 Word Search.cpp
#include <iostream>
#include <vector>
#include <cassert>
using namespace std;

class Solution {
private:
    int d[4][2] = {
            {-1, 0}, //up
            {0, 1},  //right
            {1, 0},  //down
            {0, -1}  //left
    };

    int m,n;
    vector<vector<bool>> visited;

    bool inArea(int x, int y){
        return x >= 0 && x < m && y >= 0 && y < n;
    }
    //search word[index..word.size()), from board[startx][starty]
    bool searchWord(const vector<vector<char>>& board, const string& word, int index, int startx, int starty){
        if(index == word.size()-1){
            return word[index] == board[startx][starty];
        }

        if( board[startx][starty] == word[index]){
            visited[startx][starty] = true;
            // start from startx, starty
            // search on four directions
            for(int i = 0; i < 4; i++){
                int newx = startx + d[i][0];
                int newy = starty + d[i][1];

                //check validation of newx & newy
                //check if has visited
                if(inArea(newx, newy) && !visited[newx][newy]){
                    if(searchWord(board, word, index + 1, newx, newy))
                        return true;
                }


            }
            visited[startx][starty] = false;
        }

        return false;
    }

public:
    bool exist(vector<vector<char>>& board, string word) {
        //m: row
        m = board.size();
        //n: column
        assert( m > 0);
        n = board[0].size();
        visited = vector<vector<bool>>(m, vector<bool>(n, false));

        for(int i = 0; i < board.size(); i++){
            for(int j = 0; j < board[i].size(); j++){
                if(searchWord(board, word, 0, i, j))
                    return true;
            }
        }
        return false;
    }
};

int main() {
    vector<vector<char>> board;
    char a[4] = {'A', 'B', 'C', 'E'};
    char b[4] = {'S', 'F', 'C', 'S'};
    char c[4] = {'A', 'D', 'E', 'E'};
    vector<char> va(a, a+4);
    vector<char> vb(b, b+4);
    vector<char> vc(c, c+4);
    board.push_back(va);
    board.push_back(vb);
    board.push_back(vc);

    string words = "ABCCED";

    bool res = Solution().exist(board, words);
    cout << "Result is : " << res << endl;

    return 0;
}